core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -468,21 +468,20 @@ private[spark] class DAGScheduler(
 
   /** Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet */
   private def getMissingAncestorShuffleDependencies(
-      rdd: RDD[_]): ArrayStack[ShuffleDependency[_, _, _]] = {
-    val ancestors = new ArrayStack[ShuffleDependency[_, _, _]]
+      rdd: RDD[_]): ListBuffer[ShuffleDependency[_, _, _]] = {
+    val ancestors = new ListBuffer[ShuffleDependency[_, _, _]]
     val visited = new HashSet[RDD[_]]
     // We are manually maintaining a stack here to prevent StackOverflowError
     // caused by recursively visiting
-    val waitingForVisit = new ArrayStack[RDD[_]]
-    waitingForVisit.push(rdd)
+    val waitingForVisit = ListBuffer[RDD[_]](rdd)
     while (waitingForVisit.nonEmpty) {
-      val toVisit = waitingForVisit.pop()
+      val toVisit = waitingForVisit.remove(0)
       if (!visited(toVisit)) {
         visited += toVisit
         getShuffleDependencies(toVisit).foreach { shuffleDep =>
           if (!shuffleIdToMapStage.contains(shuffleDep.shuffleId)) {
-            ancestors.push(shuffleDep)
-            waitingForVisit.push(shuffleDep.rdd)
+            ancestors.+=:(shuffleDep)
+            waitingForVisit.+=:(shuffleDep.rdd)
```

Comment:
It seems returned type confuses the compiler. Since the type is not important here, you can cast it to `Unit` like:
```scala
(waitingForVisit.+=:(dependency.rdd)): Unit
```

Reply:
I should have known.. it's a right associative operator. One is meant to write `dependency.rdd +=: waitingForVisit`. I suppose it's kind of clear that it's a prepend. Is it just clearer to use `waitingForVisit.prepend(dependency.rdd)` and accept the very small overhead of varargs? I think so

Reply:
The overhead shouldn't be so much. I would prefer `prepend(...)` for readability.